design-proposals/export-import/vm-export-import.md

```diff
@@ -0,0 +1,166 @@
+# Overview
+Proposal for exporting/importing Kubevirt Virtual Machine data from/to a Kubernetes cluster.
+
+## Motivation
+The Containerized Data Importer (CDI) project is mostly focused on getting Virtual Machine data into a cluster. The DataVolume API declares import and upload operations that populate and initialize individual Persistent Volume Claims (PVCs) with external data.  Of course that data is very likely to change once the PVCs are mounted and consumed by VirtualMachines (VMs).  But there is no standard way to get that valuable data out of the cluster.  Maybe an external backup would be useful for Disaster Recovery purposes?  Or for migrating VMs between clusters?
+
+Once there is a standard API for exporting all persistent data for a VM, it follows that a new type of import operation is required, one that operates on a higher level than the existing CDI primitatives.  It should be able to parse an exported arhive and restore the Virtual Machine configuration and data.
+
+## Goals
+* Define Virtuam Machine Archive file format
+* Define VirtualMachineExport API
+* Introduce virtctl export|import vm commands
+
+## Non Goals
+* Implementation details of internal components (servers/proxies/controllers/etc)
+
+## User Stories
+* As a KubeVirt user, I would like to backup the current state of a Virtual Machine by exporting a Virtual Machine Archive
+* As a KubeVirt user, I would like to backup the state of a Virtual Machine Snapshot by exporting a Virtual Machine Archive
+* As a KubeVirt user, I would like to create a new Virtual Machine from a Virtual Machine Archive
+
+## Repos
+* [KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+## Virtual Machine Archive File Format
+
+A Vritual Machine Archive is a xz compressed tar archive with the following structure
+
+```
+.
+├── manifest
+├── resources
+│   ├── pvc-block-pvc.yaml
+│   ├── pvc-filesystem-pvc.yaml
+│   ├── pvc-virtiofs-pvc.yaml
+│   └── vm.yaml
+└── volumes
+    ├── block-pvc
+    │   └── disk.img
+    ├── filesystem-pvc
+    │   └── disk.img
+    └── virtiofs-pvc
+        ├── file1
+        ├── file2
+        └── file3
+```
+
+### Manifest
+
+The `manifest` file lists all files (excluding itself) in the archive and their SHA-1 hashes.  It can be used to ensure that corrumted data is not imported.
+
+### Resources
+
+The `resources` path contains copies of all the Kubernetes resources associaterd to the Virtual Machine.
+
+### Volumes
+
+The `volumes` path contains an entry for each exported volume.
+
+## VirtualMachineExport API
+
+The VirtualMachineExport API allows users to download a Virtual Machine Archive from a URL.
+
+### User Flow
+
+1.  Create a VirtualMachineExport resource
+2.  Wait for the `Ready` condition of the VirtualMachineExport to be `True`
+3.  Download the Virtual Machine Archive by getting the URL in `status.url`
+
+### Export a VirtualMachine
```

Comment:
should we define what happens if the VM is started during export? (e.g we create a snapshot before export starts)

Reply:
Yeah, there are a lot of ways to handle this.  Exporting a snapshot is a great idea.  But if not possible, I guess we can either block starting the vm with a webhook or kill the export pod.  Maybe that selection can be a parameter of the VirtualMachineExport

Reply:
> But if not possible, I guess we can either block starting the vm with a webhook or kill the export pod.

Would it make sense to limit export to VM snapshots initially? That prevents us from having to care about what state the VM is currently in.

Reply:
Yeah that was exactly my thought for first phase.